src/swap/defi/thorchain.js

```diff
@@ -0,0 +1,611 @@
+      }
+
+      let responseJson = await iaResponse.json()
+      const inboundAddresses = asInboundAddresses(responseJson)
+
+      responseJson = await poolResponse.json()
+      const pools = asPools(responseJson)
+
+      // Check for supported chain and asset
+      const inAddressObject = inboundAddresses.find(
+        addrObj => addrObj.halted === false && addrObj.chain === fromMainnetCode
+      )
+      if (inAddressObject == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+      const { address: thorAddress, gas_rate: gasRate } = inAddressObject
+      clog(`${fromMainnetCode}.${fromCurrencyCode} gasRate ${gasRate}`)
+
+      const outAddressObject = inboundAddresses.find(
+        addrObj => addrObj.halted === false && addrObj.chain === toMainnetCode
+      )
+      if (outAddressObject == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+      const { gas_rate: outboundGasRate } = outAddressObject
+      clog(
+        `${toMainnetCode}.${toCurrencyCode} outboundGasRate ${outboundGasRate}`
+      )
+
+      const sourcePool = pools.find(pool => {
+        const [asset] = pool.asset.split('-')
+        return asset === `${fromMainnetCode}.${fromCurrencyCode}`
+      })
+      if (sourcePool == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+      const [sourceAsset] = sourcePool.asset.split('-')
+      clog(`sourceAsset: ${sourceAsset}`)
+
+      const destPool = pools.find(pool => {
+        const [asset] = pool.asset.split('-')
+        return asset === `${toMainnetCode}.${toCurrencyCode}`
+      })
+      if (destPool == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+
+      // Add outbound fee
+      const networkFee = calcNetworkFee(
+        toMainnetCode,
+        toCurrencyCode,
+        outboundGasRate,
+        pools
+      )
+
+      // Per Thorchain specs, user pays 2x the outbound transaction networkfee
+      let feeInDestCurrency = mul(networkFee, '2')
+      clog(`feeInDestCurrency: ${feeInDestCurrency}`)
+
+      const assetInUsd = mul(feeInDestCurrency, destPool.assetPriceUSD)
+      if (lt(assetInUsd, '1')) {
+        feeInDestCurrency = div('1', destPool.assetPriceUSD, DIVIDE_PRECISION)
+        clog(`feeInDestCurrency adjusted to $1 min: ${feeInDestCurrency}`)
+      }
+
+      let calcResponse
+      if (quoteFor === 'from') {
+        calcResponse = await calcSwapFrom(
+          {
+            fromWallet,
+            fromCurrencyCode,
+            toWallet,
+            toCurrencyCode,
+            nativeAmount,
+            sourcePool,
+            destPool,
+            volatilitySpreadFinal,
+            affiliateFee,
+            feeInDestCurrency
+          },
+          clog
+        )
+      } else {
+        throw new Error('TODO: TO quoting')
```

Comment:
Ooof, but I get it. Maybe add a more user-friendly string for the red box?

Reply:
This is removed in a later commit when reverse quotes are added